### PR TITLE
Add Söderköping, Sweden via new generic service avfallsapp.se

### PR DESCRIPTION
**Pull request by dala318**
Adding a new service provider and single instance for Söderköping Sweden

Used edpevent_se as template

### CODE REVIEW

**Comment by 5ila5** [comment]
I really don't want to merge commented out code. Either use the code or remove the code. The street_address parameter is not used at all atm, but the md file does not say that the address selected in the app is used instead. I would like the source to be finished before merging it

**Comment by dala318** [comment]
Fully understand and will adapt accordingly (can still store the deactivated code on a branch in my fork).

**Comment by dala318** [comment]
I think I have fixed your comments now, please have a look.

**Comment by 5ila5** [comment]
I'd like the Address thing to work before merging, adding additional apps could be done later and adding the address things seems not to be that hard. I did some testing around and this works for my test address:


<details><summary>CLIKC TO SHOW</summary>

```py
class Source:
    def __init__(
        self,
        service_provider: str,
        api_key: str|None= None,
        street_address: str|None= None,
    ):
        self._street_address = street_address
        if not api_key  and not street_address:
            raise SourceArgumentExceptionMultiple(
                ["street_address", "api_key"],
                "You must provide either a street_address or a api_key",
            )
        self._api_key = api_key
        # Get the api url using the service provider
        self._url = SERVICE_PROVIDERS.get(service_provider.lower(), {}).get("api_url")
        if self._url is None:
            raise SourceArgumentNotFoundWithSuggestions(
                "service_provider",
                service_provider,
                SERVICE_PROVIDERS.keys(),
            )
        # Remove trailing slash from the url if present
        if self._url.endswith("/"):
            self._url = self._url[:-1]


    def _register_device(self):
        if not self._street_address: # should not happen
            raise Exception("street_address required to register device")
        uuid = uuid4().hex[:16]

        # if not self._api_key:
        registerUrl = self._url + "/register"
        params = {
            "identifier": uuid,
            "uuid": uuid,
            "platform": "android",
            "version": "2.2.0.0",
            "os_version": "14",
            "model": "sdk_gphone64_x86_64",
            "test": False,
        }
        response = requests.post(registerUrl, json=params, timeout=30, headers={"X-App-Identifier": uuid})

       

        params = {"address": self._street_address.replace(" ", "%20")}
        # Use the street address to find the full street address with the building ID
        searchUrl = self._url + "/next-pickup/search"
        # Search for the address
        response = requests.get(searchUrl, params=params, headers={"X-App-Identifier": uuid}, timeout=30)
        address_data = json.loads(response.text)
        address = None
        # Make sure the response is valid and contains data
        if address_data and len(address_data) > 0:
            addresses = [a for _, address_list in address_data.items() for a in address_list]
            # Check if the request was successful
            for  a in addresses:
                # The request can be successful but still not return any buildings at the specified address
                if a["address"].lower().replace(" ", "") == self._street_address.lower().replace(" ", ""):
                    address = a
                    break
            if not address:
                raise SourceArgumentNotFoundWithSuggestions(
                    "street_address",
                    self._street_address,
                    [a["address"] for a in addresses],
                )

        # Raise exception if all the above checks failed
        if not address:
            raise Exception(
                "street_address",
                f"Failed to find building address for: {self._street_address}",
            )

        data = {
            "plant_id": address["plant_number"],
            "address_enabled": True,
            "notification_enabled": False,
        }

        # Set the address as the active address
        response = requests.post(self._url+"/next-pickup/set-status", json=data, headers={"X-App-Identifier": uuid}, timeout=30)
        response.raise_for_status()

        self._api_key = uuid

    def fetch(self):
        if not self._api_key:
            self._register_device()
        ... # as current version
```

</details> 

- The test case should not have this not working api_key (uuid) but "!secret avfallsapp_se_api_key" so it will look in the secrets.yaml (see initial error message when running the test)
- Filtering out the address from the bin type would also be nice (e.g. `Mellangården  Storängsparken 9 -11 Hushållsavfall` should just be `Hushållsavfall`) maybe by saving the address in the register step to then remove it in the fetch step
- Maybe put an individual ICON per waste type (like nearly every other source: add a ICON_MAP at the top and get the icno with `ICON_MAP.get(bin_type)`

**Comment by dala318** [comment]
> I'd like the Address thing to work before merging, adding additional apps could be done later and adding the address things seems not to be that hard. I did some testing around and this works for my test address:

Nice, I'll try it out and incorporate it.

> * Filtering out the address from the bin type would also be nice (e.g. `Mellangården  Storängsparken 9 -11 Hushållsavfall` should just be `Hushållsavfall`) maybe by saving the address in the register step to then remove it in the fetch step

My original thinking was since you can register several addresses to your API_KEY you would need some more identifier which "Tunna 1" to what address. But if it work as above to get your own unique key from that step you will register a new key for every address, and that problem goes away.

> * Maybe put an individual ICON per waste type (like nearly every other source: add a ICON_MAP at the top and get the icno with `ICON_MAP.get(bin_type)`

Will have a look at it, but for now I only see "Tunna 1" and "Tunna 2" which both are recycling-type (each 4-bay sorted waste)

File: custom_components/waste_collection_schedule/waste_collection_schedule/source/avfallsapp_se.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            "platform": "android",
+            "version": "2.2.0.0",
+            "os_version": "14",
+            "model": "sdk_gphone64_x86_64",
+            "test": False,
+        }
+        response = requests.post(
+            registerUrl, json=params, timeout=30, headers={"X-App-Identifier": uuid}
+        )
+        # if not response.text == "1":
```
Comment (dala318) [inline custom_components/waste_collection_schedule/waste_collection_schedule/source/avfallsapp_se.py:95]: Not ready: Wanted to add a check that API_KEY registration worked as expected, this I can likely fix fir some more testing and to comparing to negative response.

File: custom_components/waste_collection_schedule/waste_collection_schedule/source/avfallsapp_se.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        response = requests.post(
+            self._url + "/next-pickup/set-status",
+            json=data,
+            headers={"X-App-Identifier": self._api_key},
+            timeout=30,
+        )
+        response.raise_for_status()
+
+    def fetch(self):
+        if not self._api_key:
```
[the remark this answers is not in the document]
Comment (dala318) [reply custom_components/waste_collection_schedule/waste_collection_schedule/source/avfallsapp_se.py:153]: Added the raise SourceArgumentRequiredWithSuggestions below which at least get the api-key back to form for the user to select. But only happens after a single address match has been achieved.

[the remark this answers is not in the document]
Comment (5ila5) [reply custom_components/waste_collection_schedule/waste_collection_schedule/source/avfallsapp_se.py:153]: There currently is no good way to store this key persistently. I want to look at custom configuration wizards in the future but haven't had time yet. It's semi persistent, it is stored in the Source instance but a new "key" is generated after a HA restart.

Your approach is an ok workaround for the time being

[the remark this answers is not in the document]
Comment (dala318) [reply custom_components/waste_collection_schedule/waste_collection_schedule/source/avfallsapp_se.py:175]: Ok, then I'll leave it as it is.

**Comment by dala318** [comment]
I think all the comments have been corrected now, please have a look.

**Comment by 5ila5** [comment]
Thanks for your contribution

I did some minor reformatting and added a note on how to use the street_address parameter using the GUI
